From 7b2c478bb0f4c39ffc15c7e36c8bdb4b0a42fcbf Mon Sep 17 00:00:00 2001
From: Natanael Copa <ncopa@alpinelinux.org>
Date: Tue, 16 Nov 2021 16:05:04 +0100
Subject: [PATCH] setup-disk: improve bootloader selection

- unify bootloader selection.
- respect previously set BOOTLOADER, even if it may not work
- fall back to u-boot for arm and aarch64 if not efi

fixes https://gitlab.alpinelinux.org/alpine/alpine-conf/-/issues/10489
---
 setup-disk.in | 43 +++++++++++++++++++------------------------
 1 file changed, 19 insertions(+), 24 deletions(-)

diff --git a/setup-disk.in b/setup-disk.in
index 2f310fc..f949949 100644
--- a/setup-disk.in
+++ b/setup-disk.in
@@ -14,12 +14,6 @@ KERNELOPTS=${KERNELOPTS:-quiet}
 # default location for mounted root
 SYSROOT=${SYSROOT:-/mnt}
 
-# machine arch
-ARCH=$(apk --print-arch)
-case "$ARCH" in
-	x86*) BOOTLOADER=${BOOTLOADER:-syslinux};;
-esac
-
 in_list() {
 	local i="$1"
 	shift
@@ -1433,8 +1427,27 @@ shift $(( $OPTIND - 1))
 
 if is_rpi; then
 	: ${BOOTLOADER:=raspberrypi-bootloader}
+	BOOTFS=vfat
+	SWAP_SIZE=0
 fi
 
+if is_efi || [ -n "$USE_EFI" ]; then
+	USE_EFI=1
+	DISKLABEL=gpt
+	BOOT_SIZE=512
+	BOOTFS=vfat
+	: ${BOOTLOADER:=grub}
+fi
+
+# machine arch
+ARCH=$(apk --print-arch)
+case "$ARCH" in
+	x86*) 		: ${BOOTLOADER:=syslinux};;
+	ppc64le)	: ${BOOTLOADER:=grub};;
+	s390x) 		: ${BOOTLOADER:=zipl};;
+	arm*|aarch64) : ${BOOTLOADER:=u-boot};;
+esac
+
 if [ -d "$1" ]; then
 	# install to given mounted root
 	[ "$BOOTLOADER" = "syslinux" ] && apk add --quiet syslinux
@@ -1541,24 +1554,6 @@ if [ $# -gt 1 ]; then
 	USE_RAID=1
 fi
 
-if is_efi || [ -n "$USE_EFI" ]; then
-	USE_EFI=1
-	DISKLABEL=gpt
-	BOOTLOADER=grub
-	BOOT_SIZE=512
-	BOOTFS=vfat
-fi
-
-if is_rpi; then
-	BOOTFS=vfat
-	SWAP_SIZE=0
-fi
-
-case "$ARCH" in
-	ppc64le) BOOTLOADER=grub;;
-	s390x) BOOTLOADER=zipl;;
-esac
-
 dmesg -n1
 
 if [ -n "$USE_LVM" ] && ! grep -w -q device-mapper /proc/misc; then
-- 
2.34.0

